<a href="https://colab.research.google.com/github/jaleelm-blip/friendly-pancake/blob/main/lab_pandas_stats.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Statistical Methods in Pandas - Lab

## Introduction

In this lab you'll get some hands-on experience using some of the key summary statistics methods in Pandas.

## Objectives
You will be able to:

- Calculate summary statistics for a series and DataFrame
- Use the `.apply()` or `.applymap()` methods to apply a function to a pandas series or DataFrame  


## Getting Started

For this lab, we'll be working with a dataset containing information on various lego datasets. You will find this dataset in the file `'lego_sets.csv'` which is under the DATA folder.   

In the cell below:

- Import Pandas and set the standard alias of `pd`
- Import the `'lego_sets.csv'` dataset
- Display the first five rows of the DataFrame to get a feel for what we'll be working with

In [8]:
# Import pandas
import pandas as pd
# Import the 'lego_sets.csv' dataset
df = pd.read_csv ('/content/lego_sets.csv')
# Print the first five rows of DataFrame
df[:5]

ages  list_price  num_reviews  piece_count  play_star_rating  \
0  6-12       29.99          2.0        277.0               4.0   
1  6-12       19.99          2.0        168.0               4.0   
2  6-12       12.99         11.0         74.0               4.3   
3   12+       99.99         23.0       1032.0               3.6   
4   12+       79.99         14.0        744.0               3.2   

                                           prod_desc  prod_id  \
0  Catapult into action and take back the eggs fr...  75823.0   
1  Launch a flying attack and rescue the eggs fro...  75822.0   
2  Chase the piggy with lightning-fast Chuck and ...  75821.0   
3  Explore the architecture of the United States ...  21030.0   
4  Recreate the Solomon R. Guggenheim Museum® wit...  21035.0   

                                      prod_long_desc review_difficulty  \
0  Use the staircase catapult to launch Red into ...           Average   
1  Pilot Pig has taken off from Bird Island with ...              Easy   
2  Pitch speedy bird Chuck against the Piggy Car....              Easy   
3  Discover the architectural secrets of the icon...           Average   
4  Discover the architectural secrets of Frank Ll...       Challenging   

                         set_name  star_rating    theme_name  val_star_rating  \
0           Bird Island Egg Heist          4.5  Angry Birds™              4.0   
1              Piggy Plane Attack          5.0  Angry Birds™              4.0   
2                Piggy Car Escape          4.3  Angry Birds™              4.1   
3  United States Capitol Building          4.6  Architecture              4.3   
4   Solomon R. Guggenheim Museum®          4.6  Architecture              4.1   

  country  
0      US  
1      US  
2      US  
3      US  
4      US

## Getting DataFrame-Level Statistics

We'll begin by getting some overall summary statistics on the dataset. There are two ways we'll get this information -- `.info()` and `.describe()`.

The `.info()` method provides us metadata on the DataFrame itself. This allows us to answer questions such as:

* What data type does each column contain?
* How many rows are in my dataset?
* How many total non-missing values does each column contain?
* How much memory does the DataFrame take up?

In the cell below, call our DataFrame's `.info()` method.

In [9]:
# Call the .info() method
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ages               12261 non-null  object 
 1   list_price         12261 non-null  float64
 2   num_reviews        10641 non-null  float64
 3   piece_count        12261 non-null  float64
 4   play_star_rating   10486 non-null  float64
 5   prod_desc          11884 non-null  object 
 6   prod_id            12261 non-null  float64
 7   prod_long_desc     12261 non-null  object 
 8   review_difficulty  10206 non-null  object 
 9   set_name           12261 non-null  object 
 10  star_rating        10641 non-null  float64
 11  theme_name         12258 non-null  object 
 12  val_star_rating    10466 non-null  float64
 13  country            12261 non-null  object 
dtypes: float64(7), object(7)
memory usage: 1.3+ MB


#### Interpreting the Results

Read the output above, and then answer the following questions:

- How many total rows are in this DataFrame?  
- How many columns contain numeric data?
- How many contain categorical data?
- Identify at least 3 columns that contain missing values.

Write your answer below this line in comment form:
________________________________________________________________________________________________________________________________

1. There are 12261 rows in this DataFrame
2. There are 7 columns containing numeric data
3. There are 7 columns containing categorical data
4. Columns missing data include num_reviews, star_rating, and review_difficulty

Whereas `.info()` provides statistics about the DataFrame itself, `.describe()` returns output containing basic summary statistics about the data contained with the DataFrame.  

In the cell below, call the DataFrame's `.describe()` method.

In [10]:
# Call the .describe() method
df.describe()

list_price   num_reviews   piece_count  play_star_rating  \
count  12261.000000  10641.000000  12261.000000      10486.000000   
mean      65.141998     16.826238    493.405921          4.337641   
std       91.980429     36.368984    825.364580          0.652051   
min        2.272400      1.000000      1.000000          1.000000   
25%       19.990000      2.000000     97.000000          4.000000   
50%       36.587800      6.000000    216.000000          4.500000   
75%       70.192200     13.000000    544.000000          4.800000   
max     1104.870000    367.000000   7541.000000          5.000000   

            prod_id   star_rating  val_star_rating  
count  1.226100e+04  10641.000000     10466.000000  
mean   5.983675e+04      4.514134         4.228960  
std    1.638115e+05      0.518865         0.660282  
min    6.300000e+02      1.800000         1.000000  
25%    2.103400e+04      4.300000         4.000000  
50%    4.206900e+04      4.700000         4.300000  
75%    7.092200e+04      5.000000         4.700000  
max    2.000431e+06      5.000000         5.000000

#### Interpreting the Results

The output contains descriptive statistics corresponding to the columns. Use these to answer the following questions:

- How much is the standard deviation for `piece count`?
- How many pieces are in the largest lego set?
- How many in the smallest lego set? What is the median `val_star_rating`?

________________________________________________________________________________________________________________________________

In [11]:

#please use this space to answer the questions
std_piece_count = df['piece_count'].std()
print(std_piece_count)

max_piece_count = df['piece_count'].max()
print(max_piece_count)

min_piece_count = df['piece_count'].min()
print(min_piece_count)

median_val_star_rating = df['val_star_rating'].median()
print(median_val_star_rating)


825.364580411521
7541.0
1.0
4.3


## Getting Summary Statistics

Pandas also allows us to easily compute individual summary statistics using built-in methods.  Next, we'll get some practice using these methods.

In the cell below, compute the median value of the `star_rating` column.

In [12]:
# Calculate the median of the star_rating column (Hint: there is a median function)
median_star_rating = df['star_rating'].median()
print(median_star_rating)

4.7


Next, get a count of the total number of unique values in `play_star_rating`.

In [14]:
# Print the number of unique values in play_star_rating
unique_play_star_rating = df['play_star_rating'].nunique()
print(unique_play_star_rating)

30


Now, compute the standard deviation of the `list_price` column.

In [17]:
# Calculate the standard deviation of the list_price column. (Hint: you can use std function)
std_list_price = df['list_price'].std()
print(std_list_price)

91.9804293059252


If we bought every single lego set in this dataset, how many pieces would we have?  

> **Note**: If you truly want to answer this accurately, and are up for the challenge, remove duplicate lego-set entries before summing the pieces. That is, many of the lego sets are listed multiple times in the dataset above, depending on the country where it is being sold and other unique parameters. If you're stuck, just practice calculating the total number of pieces in the dataset for now.

In [20]:
# Total number of pieces across all unique Lego sets
df.drop_duplicates(subset='prod_id')['piece_count'].sum()

# If you're simply calculating the sum of a column
# df['piece_count'].sum()

np.float64(319071.0)

Now, let's try getting the value for the 90% quantile for all numerical columns.  Do this in the cell below.

In [21]:
# Get the 90% quantile for all numerical columns. (Hint: quantile function)
quantile_all = df.quantile(q=0.9, axis=1, numeric_only=True)
print(quantile_all)


0        30495.4
1        30429.6
2        30372.8
3         9031.2
4         8860.4
          ...   
12256    28448.2
12257    28381.8
12258    28382.6
12259    28280.0
12260    28307.8
Name: 0.9, Length: 12261, dtype: float64


## Getting Summary Statistics on Categorical Data

For obvious reasons, most of the methods we've used so far only work with numerical data -- there's no way to calculate the standard deviation of a column containing string values. However, there are some things that we can discover about columns containing categorical data.

In the cell below, print the unique values contained within the `review_difficulty` column.

In [ ]:
# Print the unique values in the review_difficulty column
df['review_difficulty'].unique()

array(['Average', 'Easy', 'Challenging', 'Very Easy', nan,
       'Very Challenging'], dtype=object)

Now, let's get the `value_counts()` for this column, to see how common each is.

In [ ]:
# Get the value_counts() of the review_difficulty column
print(df['review_difficulty'].value_counts())

# Alternatively normalized value counts
df['review_difficulty'].value_counts(normalize=True)

Easy                4236
Average             3765
Very Easy           1139
Challenging         1058
Very Challenging       8
Name: review_difficulty, dtype: int64


Easy                0.415050
Average             0.368901
Very Easy           0.111601
Challenging         0.103665
Very Challenging    0.000784
Name: review_difficulty, dtype: float64


## Using `.groupby`

For each review difficulty level, find how many legos you have. Then sort these based on their counts. (Hint: Please search for groupby in Pandas, `size()`, `count()`, `reset_index()`, and `sort_values()` functions)



As you can see, these provide us quick and easy ways to get information on columns containing categorical information.  


## Using `.applymap()`

When working with pandas DataFrames, we can quickly compute functions on the data contained by using the `.applymap()` method and passing in a lambda function.

For instance, we can use `applymap()` to return a version of the DataFrame where every value has been converted to a string.

In the cell below:

* Call the DataFrame's `.applymap()` method and pass in `lambda x: str(x)`  
* Call the new `string_df` object's `.info()` method to confirm that everything has been cast to a string

In [22]:
# Call the .applymap() method
string_df = df.applymap(lambda x: str(x))
# Call the .info() method
string_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ages               12261 non-null  object
 1   list_price         12261 non-null  object
 2   num_reviews        12261 non-null  object
 3   piece_count        12261 non-null  object
 4   play_star_rating   12261 non-null  object
 5   prod_desc          12261 non-null  object
 6   prod_id            12261 non-null  object
 7   prod_long_desc     12261 non-null  object
 8   review_difficulty  12261 non-null  object
 9   set_name           12261 non-null  object
 10  star_rating        12261 non-null  object
 11  theme_name         12261 non-null  object
 12  val_star_rating    12261 non-null  object
 13  country            12261 non-null  object
dtypes: object(14)
memory usage: 1.3+ MB


/tmp/ipython-input-2146133064.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_df = df.applymap(lambda x: str(x))


Note that everything -- even the `NaN` values, have been cast to a string in the example above.

Note that for Pandas Series objects (such as a single column in a DataFrame), we can do the same thing using the `.apply()` method.  

This is just one example of how we can quickly compute custom functions on our DataFrame

## Summary

In this lab, we learned how to:

* Use the `df.describe()` and `df.info()` summary statistics methods
* Use built-in Pandas methods for calculating summary statistics
* Apply a function to every element in a DataFrame